In [2]:
import numpy as np

In [3]:
def sigmoid(x):
    return 1/(1+np.exp(-x))
def sigmoid_derative(x):
    return x*(1-x)

Sigmoid takes any real number and squashes it into the range (0, 1). It's shaped like an "S" — very negative numbers go to ~0, very positive numbers go to ~1, and 0 maps to exactly 0.5.

Example: sigmoid(0.6) ≈ 0.6457, sigmoid(-3) ≈ 0.047, sigmoid(3) ≈ 0.953.

The derivative of sigmoid, but written in a shortcut form. Normally the derivative of sigmoid(x) is sigmoid(x) * (1 - sigmoid(x)). Here, the function assumes the input x is already the sigmoid output (i.e., you call sigmoid_derivative(guess), not sigmoid_derivative(sum_value)).


In [4]:
input = np.array([1.0,0.0,1.0])
target_output= 1.0
weight = np.array([0.5, -0.2, 0.1])
learning_rate = 0.1


Three inputs, one per weight.
We want the neuron to eventually output 1.0 for this input pattern.
Weights start at essentially random guesses.
learning_rate controls how big a correction to make after each mistake — small steps prevent overshooting.

In [6]:
#forward pass
sum_value = np.dot(input, weight)
sum_value

np.float64(0.6)


np.dot(inputs, weights) computes a weighted sum:
1.0×0.5+0.0×(−0.2)+1.0×0.1=0.6
Then squash it: guess=sigmoid(0.6)≈0.6457
So the network currently predicts 0.6457, but we wanted 1.0. Not terrible, but off.

In [8]:
guess = sigmoid(sum_value)
guess

np.float64(0.6456563062257954)

In [131]:
#backward pass
error = target_output - guess
print(f"erro is : ",error)
blame = error *sigmoid_derative(guess)
print(f"Blame: ", blame)
weight_adjustments = np.dot(input,blame)*learning_rate
print(f"New weights: ,",weight_adjustments)
weight += weight_adjustments
print(weight)

erro is :  0.3543436937742046
Blame:  0.08106825284073778
New weights: , [0.00810683 0.         0.00810683]
[ 1.48903268 -0.2         1.08903268]


Step by step with real numbers:

error = 1.0 - 0.6457 = 0.3543 — how far off we were.
sigmoid_derivative(guess) = 0.6457 × (1 - 0.6457) = 0.2288 — this measures how sensitive the output is to change right now. If guess were near 0 or 1 (very confident), this value shrinks toward 0, meaning "don't bother adjusting much, the neuron is already saturated." This is the classic sigmoid problem, small gradients when confident.
blame = 0.3543 × 0.2288 ≈ 0.0811 — this combines "how wrong we were" with "how much wiggle room there is to fix it."
weight_adjustments = inputs × blame × learning_rate:
weight 1 (input=1): 1 × 0.0811 × 0.1 ≈ 0.0081
weight 2 (input=0): 0 × 0.0811 × 0.1 = 0
weight 3 (input=1): 1 × 0.0811 × 0.1 ≈ 0.0081

Notice the comment is correct: inputs that were 0 get zero blame/adjustment, since they contributed nothing to the guess, there's nothing to credit or blame them for. Only weights connected to active (1) inputs get nudged.

New weights ≈ [0.5081, -0.2, 0.1081]

So weight 1 and weight 3 both increased slightly (pushing the guess higher, toward the target of 1), and weight 2 stayed exactly the same.